chainstate/src/detail/chainstateref/mod.rs

```diff
@@ -591,36 +597,54 @@ impl<'a, S: BlockchainStorageRead, V: TransactionVerificationStrategy> Chainstat
+                    Err(CheckBlockTransactionsError::InvalidDecommissionMaturityType(tx_id))
+                }
+            }
+        };
         for tx in block.transactions() {
-            tx.outputs().iter().try_for_each(|output| {
-                match output {
-                    TxOutput::Transfer(_, _)
-                    | TxOutput::Burn(_)
-                    | TxOutput::CreateStakePool(_)
-                    | TxOutput::ProduceBlockFromStake(_, _)
-                    | TxOutput::LockThenTransfer(_, _, _) => {}
-                    TxOutput::DecommissionPool(_, _, _, timelock) => match timelock {
-                        OutputTimeLock::ForBlockCount(c) => {
-                            let cs: i64 = (*c).try_into().map_err(|_| {
-                                CheckBlockTransactionsError::InvalidDecommissionMaturityDistanceValue(tx.transaction().get_id(), *c)
-                            })?;
-                            let given = BlockDistance::new(cs);
-                            let required = self.chain_config
-                                .as_ref()
-                                .decommission_pool_maturity_distance(block_height);
-                            ensure!(
-                                given >= required,
-                                CheckBlockTransactionsError::InvalidDecommissionMaturityDistance(tx.transaction().get_id(), given, required)
-                            );
-                        }
-                        OutputTimeLock::UntilHeight(_)
-                        | OutputTimeLock::UntilTime(_)
-                        | OutputTimeLock::ForSeconds(_) => {
-                            return Err(CheckBlockTransactionsError::InvalidDecommissionMaturityType(tx.transaction().get_id()));
-                        }
-                    },
-                };
-                Ok(())
+            tx.outputs().iter().try_for_each(|output| match output {
+                TxOutput::Transfer(_, _)
+                | TxOutput::Burn(_)
+                | TxOutput::CreateStakePool(_)
+                | TxOutput::ProduceBlockFromStake(_, _)
+                | TxOutput::LockThenTransfer(_, _, _)
+                | TxOutput::CreateDelegationId(_, _)
+                | TxOutput::DelegateStaking(_, _) => Ok(()),
+                TxOutput::SpendShareFromDelegation(_, _, _, timelock) => {
+                    let required =
+                        self.chain_config.as_ref().spend_share_maturity_distance(block_height);
+                    check(timelock, required, tx.transaction().get_id())
+                }
+                TxOutput::DecommissionPool(_, _, _, timelock) => {
+                    let required = self
+                        .chain_config
+                        .as_ref()
+                        .decommission_pool_maturity_distance(block_height);
+                    check(timelock, required, tx.transaction().get_id())
+                }
```

Comment:
Seems this checks the time lock field adheres to various policies imposed on it by the consensus, not that the actual time lock is satisfied. It confused me a bit at first. Maybe a more expressive name for the function would be better. Something like `check_time_lock_values` or `check_time_lock_minimal_distances` or something like that. Not that I have a great suggestion...

Reply:
renamed to `check_output_maturity_setting`